src/core/event/group.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2001-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "config.h"
+#include "group.h"
+
+#include "dev/ring.h"
+#include "event/event_handler_manager_local.h"
+#include "sock/sockinfo_tcp.h"
+
+#define MODULE_NAME "group:"
+
+#define grp_logpanic __log_panic
+#define grp_logerr   __log_err
+#define grp_logwarn  __log_warn
+#define grp_loginfo  __log_info
+#define grp_logdbg   __log_dbg
+
+xlio_group::xlio_group(const struct xlio_group_attr *attr)
+    : m_socket_event_cb(attr->socket_event_cb)
+    , m_socket_comp_cb(attr->socket_comp_cb)
+    , m_socket_rx_cb(attr->socket_rx_cb)
+    , m_flags(attr->flags)
+{
+    m_rings.reserve(2);
+    m_event_handler = new event_handler_manager_local();
+    m_tcp_timers = new tcp_timers_collection(safe_mce_sys().tcp_timer_resolution_msec,
+                                             safe_mce_sys().tcp_timer_resolution_msec);
+    m_tcp_timers->set_group(this);
+}
+
+xlio_group::~xlio_group()
+{
+    delete m_tcp_timers;
+    delete m_event_handler;
+}
+
+void xlio_group::poll()
+{
+    for (ring *rng : m_rings) {
+        uint64_t sn;
+        rng->poll_and_process_element_tx(&sn);
+        sn = 0;
+        rng->poll_and_process_element_rx(&sn);
+    }
+    m_event_handler->do_tasks();
+}
+
+void xlio_group::add_dirty_socket(sockinfo_tcp *si)
+{
+    if (m_flags & XLIO_GROUP_FLAG_DIRTY) {
+        m_dirty_sockets.push_back(si);
```

Comment:
I wonder if we want to reserve some initial space in the constructor in case XLIO_GROUP_FLAG_DIRTY is used....

Reply:
I relied on the warmup phase here. Groups should be long lived, so after few iterations the vector should group big enough.
Plus, current implementation of the dirty sockets collection is very limited and I need to rework it somehow to make it interlive with socket level flush and flushing send().